airflow/io/fsspec.py

```diff
@@ -0,0 +1,237 @@
+        """Create an input stream for reading the contents of the file.
+
+        Args:
+            seekable: If the stream should support seek, or if it is consumed sequential.
+
+        Returns:
+            OpenFile: An fsspec compliant file-like object.
+
+        Raises:
+            FileNotFoundError: If the file does not exist.
+        """
+        try:
+            return self._fs.open(self.location, "rb")
+        except FileNotFoundError as e:
+            # To have a consistent error handling experience,
+            # make sure exception contains missing file location.
+            raise e if e.filename else FileNotFoundError(
+                errno.ENOENT, os.strerror(errno.ENOENT), self.location
+            ) from e
+
+
+class FsspecOutputFile(OutputFile):
+    """An output file implementation for the FsspecFileIO.
+
+    Args:
+        location (str): A URI to a file location.
+        fs (AbstractFileSystem): An fsspec filesystem instance.
+    """
+
+    def __init__(self, location: str, fs: AbstractFileSystem):
+        self._fs = fs
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Return the total length of the file, in bytes."""
+        object_info = self._fs.info(self.location)
+        if size := object_info.get("Size"):
+            return size
+        elif size := object_info.get("size"):
+            return size
+        raise RuntimeError(f"Cannot retrieve object info: {self.location}")
+
+    def exists(self) -> bool:
+        """Check whether the location exists."""
+        return self._fs.lexists(self.location)
+
+    def create(self, overwrite: bool = False) -> OutputStream:
+        """Create an output stream for reading the contents of the file.
+
+        Args:
+            overwrite (bool): Whether to overwrite the file if it already exists.
+
+        Returns:
+            OpenFile: An fsspec compliant file-like object.
+
+        Raises:
+            FileExistsError: If the file already exists at the location and overwrite is set to False.
+
+        Note:
+            If overwrite is set to False, a check is first performed to verify that the file does not exist.
+            This is not thread-safe and a possibility does exist that the file can be created by a concurrent
+            process after the existence check yet before the output stream is created. In such a case, the
+            default behavior will truncate the contents of the existing file when opening the output stream.
+        """
+        if not overwrite and self.exists():
+            raise FileExistsError(f"Cannot create file, file already exists: {self.location}")
+        return self._fs.open(self.location, "wb")
+
+    def to_input_file(self) -> FsspecInputFile:
+        """Return a new FsspecInputFile for the location at `self.location`."""
+        return FsspecInputFile(location=self.location, fs=self._fs)
+
+
+class FsspecFileIO(FileIO):
```

Comment:
I do not fully understand the separation between the module level `FileIO` abstract base class and this implementation. Why is this separated? (Maybe I don't understand this pattern. I'd propose to push this up onto module level)